packages/tpetra/core/test/Map/Map_Bug2431.cpp

```diff
@@ -0,0 +1,347 @@
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+*/
+
+#include "Tpetra_Map.hpp"
+#include "Tpetra_DefaultPlatform.hpp"
+
+#include <vector>
+#include <unordered_map>
+
+// Test program exhibiting github issue #2431
+// Submitted by Denis Ridzal, 3/20/18
+// Modified and augmentd by Karen Devine, 3/21/18
+
+static int me;
+
+//////////////////////////////////////////////////////////////////////////////
+// Tie-break function that assigns shared IDs to the lowest process that
+// has a copy.
+namespace {
+template <typename LO, typename GO>
+class GreedyTieBreak : 
+      public Tpetra::Details::TieBreak<LO,GO> 
+{
+public:
+  GreedyTieBreak() { }
+
+  virtual bool mayHaveSideEffects() const {
+    return true;
+  }
+
+  virtual std::size_t selectedIndex(
+    GO GID,
+    const std::vector<std::pair<int,LO> > & pid_and_lid) const
+  {
+    // always choose index of pair with smallest pid
+    auto numLids = pid_and_lid.size();
+    decltype(numLids) idx = 0;
+    auto minpid = pid_and_lid[0].first;
+    decltype(minpid) minidx = 0;
+    for (idx = 0; idx < numLids; ++idx) {
+      if (pid_and_lid[idx].first < minpid) {
+        minpid = pid_and_lid[idx].first;
+        minidx = idx;
+      }
+    }
+    return minidx;
+  }
+};
+
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// Given input IDs vecP0, vecP1, vecP2, vecP3, build a (probably overlapping)
+// map with these IDs on the respective processors P0-P3.
+// Then create one-to-one maps from the overlapping map, with and without
+// use of the tie-break function.  
+// Compare the number of unique IDs in the three maps; the test passes if
+// the number of unique IDs matches.
+
+template <typename LO, typename GO, typename NO>
+int runTest(
+  const char *message,
+  std::ostream &outStream,   // allows varying levels of output
+  Teuchos::RCP<const Teuchos::Comm<int>> &comm,
+  std::vector<GO> &vecP0,
```

Comment:
Should be const reference, not nonconst reference, for the std::vector arguments.